File: leetcode/TwoSum/TwoSum.cpp
#include <vector>
#include <unordered_map>
#include <string>
#include <iostream>

class Solution {
public:
    // Approach 1: brute force
    vector<int> twoSumBruteForce(vector<int>& nums, int target) {
        int sum=0;
        for (int i=0; i<nums.size(); ++i) {
            for (int j=i+1; j<nums.size(); j++) {
                if (nums[i]+nums[j]==target) {
                    return {i,j};
                }
            }
        }
        throw std::invalid_argument("No two sum solution");
    }
    
    // Approach 2: 2-pass hash table
    // key is number, value is index
    vector<int> twoSum2PassHashMap(vector<int>& nums, int target) {
        unordered_map<int, int> map;
        vector<int> result;
        for (int i=0; i<nums.size(); ++i) {//first pass
            map.insert({nums[i],i});
        }
        for (int i=0; i<nums.size(); ++i) {//second pass
            int numToFind = target - nums[i];
            auto temp = map.find(numToFind);
            if (temp != map.end() && temp->second != i) {
                return {map[numToFind], i};
            }
        }
        throw std::invalid_argument("No two sum solution");
    }
    
    // Approach 3: 1-pass hash table
	std::vector<int> twoSum1PassHashMap(std::vector<int>& nums, int target) {
		std::unordered_map<int, int> map;
		std::vector<int> result;
        for (int i=0; i<nums.size(); ++i) {
            int numToFind = target - nums[i];
            if (map.find(numToFind) != map.end()) {
                return {map[numToFind],i};
            }
            map[nums[i]] = i;
        }
        throw std::invalid_argument("No two sum solution");
    }
};

int main (int argc, char *argv[]) {
	Solution s = Solution();
	std::vector<int> vec = {2,7,11,15};
	int target = 9;
	std::vector<int> output = s.twoSumBruteForce(vec, target);
	for (auto el : output) {
		std::cout << el << ",";
	}

	/**
	/* io  
     */
	// std::string line;
	// std::string inputArr;
	// int target;
	// std::cin >> inputArr;
	// std::cin >> target;

}

File: leetcode/TwoSum/TwoSum.java
import java.util.Map;
import java.util.HashMap;

class Solution {
	/**
	 * Solution 1: brute force
	 */
    public int[] twoSumBruteForce(int[] nums, int target) {
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[j] == target - nums[i]) {
                    return new int[] { i, j };
                }
            }
        }
        throw new IllegalArgumentException("No two sum solution");
    }
 
	/**
	 * Solution 2: 2-pass hash table 
     */
	public int[] twoSum2PassHashMap(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            map.put(nums[i], i);
        }
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement) && map.get(complement) != i) {
                return new int[] { i, map.get(complement) };
            }
        }
        throw new IllegalArgumentException("No two sum solution");
    }

	/**
	 * Solution 3: 1-pass hash table 
     */
    public int[] twoSum1PassHashMap(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[] { map.get(complement), i };
            }
            map.put(nums[i], i);
        }
        throw new IllegalArgumentException("No two sum solution");
    }

	public static void main(String[] args) {
		Solution s = new Solution();
		String line;
		int arr[] = {2,7,11,15};
		int target = 9;
		int output[] = s.twoSum2PassHashMap(arr,target);
		for (int i=0; i<output.length; ++i) {
		    System.out.println(output[i] + ",");
		}
	}
}
